shared/src/types/keccak.rs

```diff
@@ -0,0 +1,214 @@
+//! This module is for hashing Namada types using the keccak256
+//! hash function in a way that is compatible with smart contracts
+//! on Ethereum.
+use std::convert::TryFrom;
+use std::fmt::Display;
+
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use hex::FromHex;
+use thiserror::Error;
+
+use crate::types::hash::{Hash, HASH_LENGTH};
+
+/// Errors for converting / parsing Keccak hashes
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("TEMPORARY error: {error}")]
+    Temporary { error: String },
+    #[error("Failed trying to convert slice to a hash: {0}")]
+    ConversionFailed(std::array::TryFromSliceError),
+    #[error("Failed to convert string into a hash: {0}")]
+    FromStringError(hex::FromHexError),
+}
+
+/// Represents a Keccak hash.
+#[derive(
+    Clone,
+    Debug,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    BorshSerialize,
+    BorshDeserialize,
+    BorshSchema,
+)]
+pub struct KeccakHash(pub [u8; 32]);
+
+impl Display for KeccakHash {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for byte in &self.0 {
+            write!(f, "{:02X}", byte)?;
+        }
+        Ok(())
+    }
+}
+impl From<KeccakHash> for Hash {
+    fn from(hash: KeccakHash) -> Self {
+        Hash(hash.0)
+    }
+}
+
+impl From<Hash> for KeccakHash {
+    fn from(hash: Hash) -> Self {
+        KeccakHash(hash.0)
+    }
+}
+
+impl TryFrom<&[u8]> for KeccakHash {
+    type Error = Error;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        if value.len() != HASH_LENGTH {
+            return Err(Error::Temporary {
+                error: format!(
+                    "Unexpected tx hash length {}, expected {}",
+                    value.len(),
+                    HASH_LENGTH
+                ),
+            });
+        }
+        let hash: [u8; 32] =
+            TryFrom::try_from(value).map_err(Error::ConversionFailed)?;
+        Ok(KeccakHash(hash))
+    }
+}
+
+impl TryFrom<String> for KeccakHash {
+    type Error = Error;
+
+    fn try_from(string: String) -> Result<Self, Error> {
+        let bytes: Vec<u8> =
+            Vec::from_hex(string).map_err(Error::FromStringError)?;
+        Self::try_from(bytes.as_slice())
```

Comment:
We could rely on the `impl TryFrom<&str>`
```suggestion
        string.as_str().try_into()
```